src/async_io/async_io.ml

```diff
@@ -0,0 +1,303 @@
+
+module Task_id = Id.Make ()
+
+type t =
+  { readers : (Unix.file_descr, packed_task Queue.t) Table.t
+  ; writers : (Unix.file_descr, packed_task Queue.t) Table.t
+  ; mutable to_close : Unix.file_descr list
+  ; pipe_read : Unix.file_descr
+  ; pipe_write : Unix.file_descr
+  ; mutex : Mutex.t
+  ; scheduler : (module Scheduler)
+  ; mutable running : bool
+  ; mutable interrupting : bool
+  ; pipe_buf : Bytes.t
+  }
+
+and 'a task =
+  { job : unit -> 'a
+  ; ivar : ('a, [ `Cancelled | `Exn of exn ]) result Fiber.Ivar.t
+  ; select : t
+  ; what : [ `Read | `Write ]
+  ; fd : Unix.file_descr
+  ; id : Task_id.t
+  }
+
+and packed_task = Task : _ task -> packed_task
+
+let interrupt t =
+  if not t.interrupting then (
+    assert (Unix.single_write t.pipe_write byte 0 1 = 1);
+    t.interrupting <- true)
+
+module Task = struct
+  type 'a t = 'a task
+
+  let await task = Fiber.Ivar.read task.ivar
+
+  let cancel t =
+    let* () = Fiber.return () in
+    Mutex.lock t.select.mutex;
+    let+ () =
+      Fiber.Ivar.peek t.ivar >>= function
+      | Some _ -> Fiber.return ()
+      | None ->
+        let module Scheduler = (val t.select.scheduler) in
+        let table =
+          match t.what with
+          | `Read -> t.select.readers
+          | `Write -> t.select.writers
+        in
+        let should_interrupt =
+          match Table.find table t.fd with
+          | None -> false
+          | Some q ->
+            let new_q = Queue.create () in
+            Queue.iter q ~f:(fun (Task t' as task) ->
+                if Task_id.equal t.id t'.id then Scheduler.cancel_job_started ()
+                else Queue.push new_q task);
+            if Queue.is_empty new_q then (
+              Table.remove table t.fd;
+              Queue.is_empty q)
+            else (
+              Table.add_exn table t.fd new_q;
+              Queue.length new_q <> Queue.length q)
+        in
+        if should_interrupt then interrupt t.select;
+        Fiber.Ivar.fill t.ivar (Error `Cancelled)
+    in
+    Mutex.unlock t.select.mutex
+end
+
+let drain_until_ready queue acc =
+  match Queue.pop queue with
+  | None -> acc
+  | Some (Task task) ->
+    let result = try Ok (task.job ()) with exn -> Error (`Exn exn) in
+    Fiber.Fill (task.ivar, result) :: acc
+
+let make_fills fds pipe_fd waiters init =
+  List.fold_left fds ~init:(false, init) ~f:(fun (pipe, acc) fd ->
+      if Fd.equal fd pipe_fd then (true, acc)
+      else
+        let acc =
+          match Table.find waiters fd with
+          | None -> acc
+          | Some w ->
+            let acc = drain_until_ready w acc in
+            if Queue.is_empty w then Table.remove waiters fd;
+            acc
+        in
+        (pipe, acc))
+
+let rec drain_pipe pipe buf =
+  (* a single read should be enough, but just in case *)
+  match Unix.read pipe buf 0 (Bytes.length buf) with
+  | 0 -> ()
+  | _ -> drain_pipe pipe buf
+  | exception Unix.Unix_error (Unix.EAGAIN, _, _) -> ()
```

Comment:
```suggestion
let rec drain_pipe pipe buf =
  (* a single read should be enough, but just in case *)
  match Unix.read pipe buf 0 (Bytes.length buf) with
  | 0 -> ()
  | _ -> drain_pipe pipe buf
  | exception Unix.Unix_error (Unix.EAGAIN, _, _) -> ()
```
This function will block forever if the pipe is blocking, right? You can use https://learn.microsoft.com/en-us/windows/win32/api/namedpipeapi/nf-namedpipeapi-peeknamedpipe to check if a pipe has bytes available.

Reply:
Good call. I removed the recursive read. It's not really needed.